middleware/grpc_instrumentation.go

```diff
@@ -6,79 +6,76 @@ package middleware
 
 import (
 	"context"
+	"errors"
 	"io"
 	"strconv"
 	"time"
 
 	"github.com/prometheus/client_golang/prometheus"
 	"go.uber.org/atomic"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/metadata"
 
 	"github.com/grafana/dskit/grpcutil"
-	"github.com/grafana/dskit/httpgrpc"
 	"github.com/grafana/dskit/instrument"
 )
 
-func observe(ctx context.Context, hist *prometheus.HistogramVec, method string, err error, duration time.Duration) {
-	respStatus := "success"
-	if err != nil {
-		if errResp, ok := httpgrpc.HTTPResponseFromError(err); ok {
-			respStatus = strconv.Itoa(int(errResp.Code))
-		} else if grpcutil.IsCanceled(err) {
-			respStatus = "cancel"
-		} else {
-			respStatus = "error"
-		}
-	}
-	instrument.ObserveWithExemplar(ctx, hist.WithLabelValues(gRPC, method, respStatus, "false"), duration.Seconds())
+func observe(ctx context.Context, hist *prometheus.HistogramVec, method string, err error, duration time.Duration, instrumentationLabelOptions ...InstrumentationLabelOption) {
+	instrumentationLabel := ApplyInstrumentationLabelOptions(false, instrumentationLabelOptions...)
```

Comment:
Let's not do this on every call to `observe`. We can compute whether grpc codes are wanted or not just once per each interceptor, and pass that value to `observe`.

Reply:
I have fixed that in all the places where `ApplyInstrumentationLabelOptions` was called.